dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -117,12 +120,22 @@
         this.retryOnTimeout = retryOnTimeout;
         this.scheduler = instrument(scheduler, metrics);
         this.jitter = jitter;
+        // we want to give users helpful stacktraces without paying the cost of creating one for every call
+        this.shouldRecordStacktrace = Caffeine.newBuilder()
+                .maximumSize(1000)
+                .expireAfterAccess(Duration.ofMinutes(1))
+                .build(); // not plumbing in a clock because this is purely informational
     }
 
     @Override
     public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
         if (isRetryable(request)) {
-            return new RetryingCallback(endpoint, request).execute();
+            Optional<SafeRuntimeException> exceptionForStacktrace =
+                    shouldRecordStacktrace.getIfPresent(endpoint) != null
+                            ? Optional.of(new SafeRuntimeException("Exception for stacktrace"))
+                            : Optional.empty();
```

Comment:
I'm worried about the implications for high-load services running in environments without the best routing hardware, this could further tank performance. Let's chat through it at tomorrows standup :-)